File: config/scouting/2026/rankings_2026.py
'''
python rankings_2026.py

    --event         tba/frc event key
    --csv           filename of tpw data
    --baseFilePath  base filesystem path

stores rankings in json file:

    filename: [event]-rankings.json

caches parsed data to json file:

    filename:   parsed_tpw_data_[event].json
'''


import numpy as np
from collections import OrderedDict
import json
import os
import math
import sys
import csv
import pandas as pd

rawArgs = sys.argv[1:]
args = {}
for i in range(len(rawArgs)):
    if rawArgs[i] == "--event" and "event" not in args:
        args["event"] = rawArgs[i + 1]
        i += 1
    elif rawArgs[i] == "--csv" and "csv" not in args:
        args["csv"] = rawArgs[i + 1]
        i += 1
    elif rawArgs[i] == "--baseFilePath" and "baseFilePath" not in args:
        args["baseFilePath"] = rawArgs[i + 1]
        i += 1

event = args["event"]
base = args["baseFilePath"]
tpw_csv = args["csv"]

def avg(data):
    if data != []:
        data = np.array([data])
        return np.mean(data)
    else:
        return 0

def std(data):
    if data != []:
        data = np.array([data])
        return np.std(data)
    else:
        return 0

def max(data):
    if data != []:
        data = np.array([data])
        return np.max(data)
    else:
        return 0

def min(data):
    if data != []:
        data = np.array([data])
        return np.min(data)
    else:
        return 0

tpw_path = base + tpw_csv

def getData():
    team_data = OrderedDict()
    data_length = 0

    if os.path.exists(tpw_path):
        with open(tpw_path, "r") as file:
            TPW_data = csv.DictReader(file)
            for x in TPW_data:
                data_length += 1
                if x['team'] not in team_data:
                    team_data[x['team']] = [x]
                else:
                    team_data[x['team']].append(x)
    else:
        raise Exception("Could not find TPW file")

    parsed_tpw_data = OrderedDict()
    for team, dict in team_data.items():
        afgps = list()
        tfgps = list()
        afgpts = {}
        tfgpts = {}
        l1climbs = list()
        egcpts = list()  # endgame climb points
        defe = list()
        speed = list()
        driver = list()
        stab = list()
        inta = list()
        uptime = list()
        avg_auto_points = list()
        avg_tele_points = list()
        matches = {}

        for x in dict:
            auto_fuel_pieces = x['auto fuel scoring'][1:len(x['auto fuel scoring']) - 1].split(", ")
            tele_fuel_pieces = x['teleop fuel scoring'][1:len(x['teleop fuel scoring']) - 1].split(", ")
            game_pieces = auto_fuel_pieces + tele_fuel_pieces
            afgps.append(auto_fuel_pieces)
            tfgps.append(tele_fuel_pieces)
            l1climbs.append(x.get('l1 climb', '').lower() == 'true' or x.get('l1 climb', '') == True)

            c_lev = int(x['climb level'])
            if c_lev == 0:
                egcpts.append(0)
            elif c_lev == 1:
                egcpts.append(10)
            elif c_lev == 2:
                egcpts.append(20)
            elif c_lev >= 3:
                egcpts.append(30)

            try:
                defe.append(int(x["defense skill"]))
                speed.append(int(x["speed"]))
                stab.append(int(x["stability"]))
                inta.append(int(x["intake consistency"]))
                driver.append(int(x["driver skill"]))
                uptime.append(153000 - int(x["brick time"]))
            except:
                defe.append(3)
                speed.append(3)
                stab.append(3)
                inta.append(3)
                driver.append(3)
                uptime.append(100)

            try:
                matches[x['match']][(x[''])] = game_pieces
            except:
                matches[x['match']] = {x['']: game_pieces}

        for i in range(len(afgps)):
            afgpts[i] = 0
            for j in range(len(afgps[i])):
                val = afgps[i][j]
                if val == "fsa":
                    afgpts[i] = afgpts.get(i, 0) + 1
                else:
                    afgpts[i] = afgpts.get(i, 0) + 0
            if l1climbs[i]:
                afgpts[i] = afgpts.get(i, 0) + 15
            avg_auto_points.append(afgpts[i])
        for i in range(len(tfgps)):
            tfgpts[i] = 0
            for j in range(len(tfgps[i])):
                val = tfgps[i][j]
                if val == "fsa":
                    tfgpts[i] = tfgpts.get(i, 0) + 1
                elif val == "fp":
                    tfgpts[i] = tfgpts.get(i, 0) + 0
                else:
                    tfgpts[i] = tfgpts.get(i, 0) + 0
            avg_tele_points.append(tfgpts[i])

        data_tpw = OrderedDict()
        data_tpw['avg-tele'] = avg(avg_tele_points)
        data_tpw['avg-auto'] = avg(avg_auto_points)
        data_tpw['avg-climb'] = avg(egcpts)
        data_tpw['avg-def'] = avg(defe)
        data_tpw['avg-driv'] = avg(driver)
        data_tpw['avg-speed'] = avg(speed)
        data_tpw['avg-stab'] = avg(stab)
        data_tpw['avg-inta'] = avg(inta)
        data_tpw['avg-upt'] = avg(uptime)
        data_tpw['matches'] = matches
        data_tpw['tpw-std'] = std(avg_auto_points) + std(avg_tele_points) + std(egcpts)
        data_tpw["tpw-score"] = data_tpw['avg-auto'] + data_tpw['avg-tele'] + data_tpw['avg-climb']
        parsed_tpw_data[team] = data_tpw

    with open(base + 'parsed_tpw_data_'+event+'.json', 'w') as f:
        f.write(json.dumps({'lines': data_length, 'data': parsed_tpw_data}, default=int))
        f.close()
    return parsed_tpw_data

def getDataLength():
    data_length = 0
    if os.path.exists(tpw_path):
        with open(tpw_path, "r") as file:
            TPW_data = csv.DictReader(file)
            for x in TPW_data:
                data_length += 1
    else:
        raise Exception("Could not find TPW file")

    return data_length


if os.path.exists(base + 'parsed_tpw_data_'+event+'.json'):
    with open(base + 'parsed_tpw_data_'+event+'.json') as f:
        loaded = json.loads(f.read())
        if loaded['lines'] == getDataLength():
            parsed_tpw_data = loaded['data']
            f.close()
        else:
            f.close()
            parsed_tpw_data = getData()
else:
    parsed_tpw_data = getData()

for team, dict in parsed_tpw_data.items():
    parsed_tpw_data[team]['r-score'] = parsed_tpw_data[team]["tpw-score"] - parsed_tpw_data[team]["tpw-std"] + parsed_tpw_data[team]["avg-driv"] + parsed_tpw_data[team]["avg-speed"] + parsed_tpw_data[team]["avg-stab"] + parsed_tpw_data[team]["avg-inta"]

sorted_dict = OrderedDict(sorted(parsed_tpw_data.items(), key=lambda x: x[1]["r-score"]))
public_dict = OrderedDict()

for team, dict in sorted_dict.items():
    public_dict[team] = {"off-score": dict["r-score"], "def-score": dict["avg-def"]}

with open(base + event + "-rankings.json", "w") as f:
    json.dump(public_dict, f)

File: config/scouting/2026/rankings_2026.ts
import { parsedRow } from ".";

interface teamData {
    [key: string]: any[];
}

interface parsedTPWData {
    [team: string]: {
        "avg-tele": number;
        "avg-auto": number;
        "avg-climb": number;
        "avg-def": number;
        "avg-driv": number;
        "avg-speed": number;
        "avg-stab": number;
        "avg-inta": number;
        "avg-upt": number;
        matches: any;
        "tpw-std": number;
        "tpw-score": number;
        "r-score"?: number;
    };
}

interface rankings {
    [team: string]: {
        "off-score": number;
        "def-score": number;
    };
}

// sum and average taken from https://gist.github.com/dggluz/365527824f9f521055baa3532b1d46e7
const sum = (numbers: number[]) =>
    numbers.reduce((total, aNumber) => total + aNumber, 0);
const avg = (numbers: number[]) => sum(numbers) / numbers.length;

// std taken from https://decipher.dev/30-seconds-of-typescript/docs/standardDeviation/
const std = (arr) => {
    const mean = arr.reduce((acc, val) => acc + val, 0) / arr.length;
    return Math.sqrt(
        arr
            .reduce((acc, val) => acc.concat((val - mean) ** 2), [])
            .reduce((acc, val) => acc + val, 0) / arr.length
    );
};

function getData(data): any {
    let team_data: teamData = {};
    if (!data) {
        throw new Error("No data given to getData()");
    }
    for (let x of data) {
        if (team_data[x["team"]] == null) {
            team_data[x["team"]] = [x];
        } else {
            team_data[x["team"]].push(x);
        }
    }
    let parsed_tpw_data: parsedTPWData = {};
    for (let team in team_data) {
        let afgps = [];
        let tfgps = [];
        let afgpts = {};
        let tfgpts = {};
        let l1climbs = [];
        let egcpts = [];
        let defe = [];
        let speed = [];
        let driver = [];
        let stab = [];
        let inta = [];
        let uptime = [];
        let avg_auto_points = [];
        let avg_tele_points = [];
        let matches = {};
        for (let x of team_data[team]) {
            let auto_fuel_pieces = x["auto fuel scoring"]
                .slice(1, x["auto fuel scoring"].length - 1)
                .split(", ");
            let tele_fuel_pieces = x["teleop fuel scoring"]
                .slice(1, x["teleop fuel scoring"].length - 1)
                .split(", ");
            let game_pieces = auto_fuel_pieces.concat(tele_fuel_pieces);
            afgps.push(auto_fuel_pieces);
            tfgps.push(tele_fuel_pieces);
            l1climbs.push(x["l1 climb"] === true || x["l1 climb"] === "true");
            let climb_lev = parseInt(x["climb level"]);
            if (climb_lev == 0) {
                egcpts.push(0);
            } else if (climb_lev == 1) {
                egcpts.push(10);
            } else if (climb_lev == 2) {
                egcpts.push(20);
            } else if (climb_lev >= 3) {
                egcpts.push(30);
            }
            try {
                defe.push(parseInt(x["defense skill"]));
                speed.push(parseInt(x["speed"]));
                stab.push(parseInt(x["stability"]));
                inta.push(parseInt(x["intake consistency"]));
                driver.push(parseInt(x["driver skill"]));
                uptime.push(153000 - parseInt(x["brick time"]));
            } catch {
                defe.push(3);
                speed.push(3);
                stab.push(3);
                inta.push(3);
                driver.push(3);
                uptime.push(100);
            }
            try {
                matches[x["match"]][x[""]] = game_pieces;
            } catch {
                matches[x["match"]] = { [x[""]]: game_pieces };
            }
        }
        for (let i = 0; i < afgps.length; ++i) {
            for (let j = 0; j < afgps[i].length; ++j) {
                let val = afgps[i][j];
                if (val == "fsa") {
                    afgpts[i] = (afgpts[i] || 0) + 1;
                } else {
                    afgpts[i] = (afgpts[i] || 0) + 0;
                }
            }
            if (l1climbs[i]) {
                afgpts[i] = (afgpts[i] || 0) + 15;
            }
            avg_auto_points.push(afgpts[i]);
        }
        for (let i = 0; i < tfgps.length; ++i) {
            for (let j = 0; j < tfgps[i].length; ++j) {
                let val = tfgps[i][j];
                if (val == "fsa") {
                    tfgpts[i] = (tfgpts[i] || 0) + 1;
                } else if (val == "fp") {
                    tfgpts[i] = (tfgpts[i] || 0) + 0;
                } else {
                    tfgpts[i] = (tfgpts[i] || 0) + 0;
                }
            }
            avg_tele_points.push(tfgpts[i]);
        }
        let data_tpw: parsedTPWData[string] = {
            "avg-tele": avg(avg_tele_points),
            "avg-auto": avg(avg_auto_points),
            "avg-climb": avg(egcpts),
            "avg-def": avg(defe),
            "avg-driv": avg(driver),
            "avg-speed": avg(speed),
            "avg-stab": avg(stab),
            "avg-inta": avg(inta),
            "avg-upt": avg(uptime),
            matches: matches,
            "tpw-std":
                std(avg_auto_points) + std(avg_tele_points) + std(egcpts),
            "tpw-score": 0
        };
        data_tpw["tpw-score"] =
            data_tpw["avg-auto"] + data_tpw["avg-tele"] + data_tpw["avg-climb"];
        parsed_tpw_data[team] = data_tpw;
    }
    return parsed_tpw_data;
}

export function computeRankings(data: parsedRow[]): rankings {
    const parsed_data = getData(data);
    for (let team in parsed_data)
        parsed_data[team]["r-score"] =
            parsed_data[team]["tpw-score"] -
            parsed_data[team]["tpw-std"] +
            parsed_data[team]["avg-driv"] +
            parsed_data[team]["avg-speed"] +
            parsed_data[team]["avg-stab"] +
            parsed_data[team]["avg-inta"];
    const sorted = Object.entries(parsed_data).sort(
        (a, b) => (b[1]["r-score"] || 0) - (a[1]["r-score"] || 0)
    );
    const pranks: rankings = {};
    for (const [team, data] of sorted) {
        pranks[team] = {
            "off-score": data["r-score"] || 0,
            "def-score": data["avg-def"]
        };
    }
    return pranks;
}
